File: java/src/bowling/GameTest.java
package bowling;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class GameTest {

    private Game game;

    @Before
    public void setup() {
        game = new Game();
    }

    @Test
    public void gutterGame() throws Exception {
        rollMany(20, 0);
        assertEquals(0, game.score());
    }

    @Test
    public void onePin() throws Exception {
        game.roll(1);
        rollMany(19, 0);
        assertEquals(1, game.score());
    }

    @Test
    public void allOnes() throws Exception {
        rollMany(20, 1);
        assertEquals(20, game.score());
    }

    @Test
    public void spare() throws Exception {
        game.roll(9);
        game.roll(1);
        game.roll(4);
        rollMany(17, 0);
        assertEquals(18, game.score());
    }

    @Test
    public void strike() throws Exception {
        game.roll(10);
        game.roll(4);
        game.roll(5);
        rollMany(16, 0);
        assertEquals(28, game.score());
    }

    @Test
    public void perfectGame() throws Exception {
        rollMany(12, 10);
        assertEquals(300, game.score());
    }

    @Test
    public void allSpare() throws Exception {
        rollMany(21, 5);
        assertEquals(150, game.score());
    }

    @Test
    public void heartBreaker() throws Exception {
        rollMany(11, 10);
        game.roll(9);
        assertEquals(299, game.score());
    }

    @Test
    public void frameObserverAllOnes() throws Exception {
        rollMany(20, 1);
        MockFrameObserver observer = new MockFrameObserver();
        game.score(observer);
        assertEquals(10, observer.frameScores.size());
        assertEquals(2, (int) observer.frameScores.get(0));
        assertEquals(4, (int) observer.frameScores.get(1));
        assertEquals(6, (int) observer.frameScores.get(2));
        assertEquals(8, (int) observer.frameScores.get(3));
        assertEquals(10, (int) observer.frameScores.get(4));
        assertEquals(12, (int) observer.frameScores.get(5));
        assertEquals(14, (int) observer.frameScores.get(6));
        assertEquals(16, (int) observer.frameScores.get(7));
        assertEquals(18, (int) observer.frameScores.get(8));
        assertEquals(20, (int) observer.frameScores.get(9));
    }

    @Test
    public void frameObserverPerfectGame() throws Exception {
        rollMany(12, 10);
        MockFrameObserver observer = new MockFrameObserver();
        game.score(observer);
        assertEquals(10, observer.frameScores.size());
        assertEquals(30, (int) observer.frameScores.get(0));
        assertEquals(60, (int) observer.frameScores.get(1));
        assertEquals(90, (int) observer.frameScores.get(2));
        assertEquals(120, (int) observer.frameScores.get(3));
        assertEquals(150, (int) observer.frameScores.get(4));
        assertEquals(180, (int) observer.frameScores.get(5));
        assertEquals(210, (int) observer.frameScores.get(6));
        assertEquals(240, (int) observer.frameScores.get(7));
        assertEquals(270, (int) observer.frameScores.get(8));
        assertEquals(300, (int) observer.frameScores.get(9));
    }

    private void rollMany(int rolls, int pins) {
        for(int i = 0; i < rolls; i++) {
            game.roll(pins);
        }
    }
}

File: typescript/src/bowling/game.test.ts
import { FrameObserver } from './frame_observer';
import { Game } from './game';
import { MockFrameObserver } from './mock_frame_observer';

describe('Game', () => {
    let game: Game;

    beforeEach(() => {
        game = new Game();
    });

    const rollMany = (rolls: number, pins: number): void => {
        for (let i = 0; i < rolls; i++) {
            game.roll(pins);
        }
    };

    test('gutter game scores 0', () => {
        rollMany(20, 0);
        expect(game.score()).toBe(0);
    });

    test('one pin scores 1', () => {
        game.roll(1);
        rollMany(19, 0);
        expect(game.score()).toBe(1);
    });

    test('all ones scores 20', () => {
        rollMany(20, 1);
        expect(game.score()).toBe(20);
    });

    test('spare adds next roll', () => {
        game.roll(9);
        game.roll(1); // Spare
        game.roll(4);
        rollMany(17, 0);
        expect(game.score()).toBe(18);
    });

    test('strike adds next two rolls', () => {
        game.roll(10); // Strike
        game.roll(4);
        game.roll(5);
        rollMany(16, 0);
        expect(game.score()).toBe(28);
    });

    test('perfect game', () => {
        rollMany(12, 10);
        expect(game.score()).toBe(300);
    });

    test('all spares', () => {
        rollMany(21, 5);
        expect(game.score()).toBe(150);
    });

    test('heart breaker', () => {
        rollMany(11, 10);
        game.roll(9);
        expect(game.score()).toBe(299);
    });

    test('frame observer - all ones', () => {
        rollMany(20, 1);
        const observer: MockFrameObserver = new MockFrameObserver();
        game.score(observer);
        expect(observer.frameScores.length).toBe(10);
        expect(observer.frameScores[0]).toBe(2);
        expect(observer.frameScores[1]).toBe(4);
        expect(observer.frameScores[2]).toBe(6);
        expect(observer.frameScores[3]).toBe(8);
        expect(observer.frameScores[4]).toBe(10);
        expect(observer.frameScores[5]).toBe(12);
        expect(observer.frameScores[6]).toBe(14);
        expect(observer.frameScores[7]).toBe(16);
        expect(observer.frameScores[8]).toBe(18);
        expect(observer.frameScores[9]).toBe(20);
    });

    test('frame observer - perfect game', () => {
        rollMany(12, 10);
        const observer: MockFrameObserver = new MockFrameObserver();
        game.score(observer);
        expect(observer.frameScores.length).toBe(10);
        expect(observer.frameScores[0]).toBe(30);
        expect(observer.frameScores[1]).toBe(60);
        expect(observer.frameScores[2]).toBe(90);
        expect(observer.frameScores[3]).toBe(120);
        expect(observer.frameScores[4]).toBe(150);
        expect(observer.frameScores[5]).toBe(180);
        expect(observer.frameScores[6]).toBe(210);
        expect(observer.frameScores[7]).toBe(240);
        expect(observer.frameScores[8]).toBe(270);
        expect(observer.frameScores[9]).toBe(300);
    });

});